{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8873cd30",
   "metadata": {},
   "source": [
    "# This script will used OSCA (OmicS-data-based Complex trait Analysis) for analysis of variance explained and mixed linear models \n",
    "\n",
    "Author: Jose Jaime Martinez-Magana\n",
    "\n",
    "Day: 15 February 2023\n",
    "\n",
    "This script will use OSCA to estimate variance explained and mixed linear models in the analysis of SES in children.\n",
    "OSCA is found in https://yanglab.westlake.edu.cn/software/osca/#Overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f1ca085",
   "metadata": {},
   "outputs": [],
   "source": [
    "# path to OSCA folder\n",
    "# /gpfs/gibbs/project/montalvo-ortiz/jjm262/programs/osca\n",
    "# this path should be change to your specific path\n",
    "# downloading OSCA\n",
    "wget https://yanglab.westlake.edu.cn/software/osca/download/osca-0.46.1-linux-x86_64.zip\n",
    "# unzip the file\n",
    "unzip osca-0.46.1-linux-x86_64.zip\n",
    "# remove zip file\n",
    "rm osca-0.46.1-linux-x86_64.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55fe4900",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first we need to create an annotation matrix fallowing the structure of annotation\n",
    "#chr   probeID    pos  probeID  strand\n",
    "\n",
    "# request computational resources\n",
    "srun --pty --mem=8G -p interactive bash\n",
    "# load environment\n",
    "module load miniconda\n",
    "# activate environment\n",
    "conda activate ewas_saliva\n",
    "\n",
    "# openning R\n",
    "R\n",
    "# loading annotation data\n",
    "library(IlluminaHumanMethylation450kanno.ilmn12.hg19)\n",
    "data(IlluminaHumanMethylation450kanno.ilmn12.hg19)\n",
    "\n",
    "# setting output path\n",
    "outp=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/annot/annot_file.txt\"\n",
    "\n",
    "# get illunina annotations\n",
    "kanno=IlluminaHumanMethylation450kanno.ilmn12.hg19::Locations\n",
    "# remove \"chr\" from chromosome\n",
    "kanno$chr=gsub(\"chr\",\"\",kanno$chr)\n",
    "# change X code to 23 and Y to 24\n",
    "kanno$chr[kanno$chr == \"X\"] = 23\n",
    "kanno$chr[kanno$chr == \"Y\"] = 24\n",
    "# transforming position and chromosome to numeric\n",
    "kanno$pos=as.numeric(kanno$pos)\n",
    "kanno$chr=as.numeric(kanno$chr)\n",
    "# ordering by chromosome and position\n",
    "kanno=kanno[order(kanno$chr, kanno$pos),]\n",
    "# changing the annotation structure of the dataframe\n",
    "kanno2=data.frame(chr=kanno$chr, probeID=rownames(kanno), pos=kanno$pos, probeID2=rownames(kanno), strand=kanno$strand)\n",
    "# write annotation dataframe\n",
    "write.table(file=outp,\n",
    "           kanno2,\n",
    "           row.names=FALSE,\n",
    "           sep=\"\\t\",\n",
    "           quote=FALSE)\n",
    "# exit R\n",
    "q()\n",
    "\n",
    "# deactivate environment\n",
    "conda deactivate\n",
    "\n",
    "# exit computaional resources\n",
    "exit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afc52ee4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# request computational resources\n",
    "# we are going to use a script that requieres multithreads to run change the --threads parameter for speeding\n",
    "srun --pty --mem=32G --threads=4 -p interactive bash\n",
    "# load environment\n",
    "module load miniconda\n",
    "# activate environment\n",
    "conda activate ewas_saliva\n",
    "\n",
    "####################################################################################\n",
    "# we developed a script for extracting the beta matrix, covariates, and phenofiles \n",
    "# from the qcdata for saliva samples and transform to osca format\n",
    "# the script is called qcdata_to_osca.Rscript\n",
    "#################################################################################################\n",
    "# content of the script\n",
    "\n",
    "#!/usr/bin/env Rscript --vanilla --slave\n",
    "##################################################################################################\n",
    "# Rscript to generate input files for osca, using the qc'ed file \n",
    "# day: 15 February 2023\n",
    "# author: Jose Jaime Martinez-Magana\n",
    "####################################################################################\n",
    "# This script uses three inputs the qcdata following this github https://github.com/martinezjaime/ewas_saliva_ses/blob/main/qc_data/probe_and_sample_quality_control.ipynb,\n",
    "# a list of samples to be included in the analysis as csv file\n",
    "\n",
    "library(optparse) \n",
    "option_list=list(\n",
    "    make_option(c(\"--file\"),\n",
    "                type=\"character\",\n",
    "                default=NULL,\n",
    "                help=\"path to *rds object with qc information\", metavar=\"character\"),\n",
    "    make_option(c(\"--out\"),\n",
    "                type=\"character\",\n",
    "                default=\"out.rds\",\n",
    "                help=\"output file name for rds [default= %default]\", metavar=\"character\"),\n",
    "    make_option(c(\"--outname\"),\n",
    "                type=\"character\",\n",
    "                default=NULL,\n",
    "                help=\"name of the rds file after performing this analysis\", metavar=\"character\"),\n",
    "    make_option(c(\"--samplelist\"),\n",
    "                type=\"character\",\n",
    "                default=NULL,\n",
    "                help=\"path to a *csv file with a list of samples to subset the analysis. This script uses a file with SampleID with Array_Sentrix structures and a header with SampleID\", metavar=\"character\"),\n",
    "    make_option(c(\"--pheno\"),\n",
    "                type=\"character\",\n",
    "                default=NULL,\n",
    "                help=\"name of the column identifying phenotype to be tested in the regression model\", metavar=\"character\"),\n",
    "    make_option(c(\"--covar\"),\n",
    "                type=\"character\",\n",
    "                default=NULL,\n",
    "                help=\"list of covariates to be included in the analysis from the phenofile separeted by ','\", metavar=\"character\")\n",
    ");\n",
    "\n",
    "opt_parser=OptionParser(option_list=option_list);\n",
    "opt=parse_args(opt_parser);\n",
    "\n",
    "if (is.null(opt$file)){\n",
    "  print_help(opt_parser)\n",
    "  stop(\"At least one argument must be supplied (input file)\", call.=FALSE)\n",
    "}\n",
    "\n",
    "########################################################################################################################\n",
    "# this part of the script will be removed from the script\n",
    "# is only for testing the script\n",
    "# create a empty list to simulate the options in optparse library\n",
    "opt=list()\n",
    "opt$file=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/qced_data_v02062023.rds\"\n",
    "opt$out=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca\"\n",
    "opt$outname=\"beta_gcta_v02152023\"\n",
    "opt$pheno=c(\"SES\")\n",
    "opt$covar=c('age,gender')\n",
    "########################################################################################################################\n",
    "\n",
    "# load minfi for package compatibility\n",
    "library(minfi)\n",
    "library(data.table)\n",
    "\n",
    "# loading rds\n",
    "paste0(\"Start analysis of data:\",Sys.time(),\"---\",\"###Analysis path[\",opt$file,\"]###\")\n",
    "file=readRDS(opt$file)\n",
    "# setting output file\n",
    "outfile=paste0(opt$out,opt$outname,sep=\"\")\n",
    "paste0(\"Input files for gcta will be saved to:\",Sys.time(),\"---\",\"###Output path[\",outfile,\"_*]###\")\n",
    "\n",
    "# loading phenotype file\n",
    "paste0(\"Loading pheno file:\",Sys.time())\n",
    "# add pheno to be tested\n",
    "pheno=opt$pheno\n",
    "# add covariates to be tested\n",
    "covar=unlist(strsplit(opt$covar,','))\n",
    "# get covars from pheno file\n",
    "covars=c(pheno,covar)\n",
    "covars=c(pheno,covar,\"SampleID\")\n",
    "# select covars from phenofile\n",
    "pheno_data=file$pheno[,colnames(file$pheno) %in% covars]\n",
    "# add SampleID as rownames\n",
    "rownames(pheno_data)=pheno_data$SampleID\n",
    "# remove SampleID column\n",
    "pheno_data$SampleID=NULL\n",
    "\n",
    "# get cells\n",
    "paste0(\"Adding cells to pheno file:\",Sys.time())\n",
    "cells=file$cell_epidish\n",
    "# transform cells to data frame\n",
    "cells2=as.data.frame(cells)\n",
    "# remove unneccesary object\n",
    "rm(cells)\n",
    "# removing cells with cero numbers in all samples\n",
    "cells3=cells2[,colSums(cells2) > 0.00]\n",
    "paste0(\"Warning: This script merges CD4T + CD8T cells to Limphocytes imputed with EpiDISH from saliva, please revised the cell distribution in your data\")\n",
    "# merging CD4T and CD8T cells\n",
    "cells3_m=cells3[,\"CD4T\", drop = F] + cells3[,\"CD8T\", drop = F]\n",
    "colnames(cells3_m)=\"Lympho\"\n",
    "cells3_m2=cbind(cells3_m, cells3[,c(\"Epi\",\"Fib\",\"B\",\"NK\",\"Mono\",\"Neutro\")])\n",
    "# add warning to say the user the cells that will be used in the analysis\n",
    "paste0(\"Using the following cells in the regression models:\",colnames(cells3_m2))\n",
    "# merge pheno_data with cells\n",
    "pheno_data=cbind(pheno_data,cells3_m2)\n",
    "\n",
    "# subset samples if sample list is added\n",
    "# subset phenofile\n",
    "# here we used is.null, to ask if there was a sample list added to the script\n",
    "if(!is.null(opt$samplelist)){\n",
    "    paste0(\"Sample filter added, phenofile will be filtered based in:\",opt$samplelist,\"---\",Sys.time())\n",
    "    sample_f=read.csv(opt$samplelist)\n",
    "    # subsetting pheno file\n",
    "    pheno_data=pheno_data[rownames(pheno_data) %in% sample_f$SampleID,]\n",
    "} else {\n",
    "    pheno_data=pheno_data\n",
    "}\n",
    "\n",
    "# subset beta matrix\n",
    "if(!is.null(opt$samplelist)){\n",
    "    paste0(\"Sample filter added, beta matrix will be filtered based in:\",opt$samplelist,\"---\",Sys.time())\n",
    "    sample_f=read.csv(opt$samplelist)\n",
    "    bvalues=file$rgsetraw_bmiq\n",
    "    bvalues=bvalues[,colnames(bvalues) %in% sample_f$SampleID]\n",
    "} else {\n",
    "    bvalues=file$rgsetraw_bmiq\n",
    "}\n",
    "\n",
    "# sample sanity check\n",
    "all_samples=is.null(pheno_data[!rownames(pheno_data) %in% colnames(bvalues),])\n",
    "if(all_samples == TRUE){\n",
    "    paste0(\"All samples in the phenofile found in the beta value matrix\")\n",
    "    bvalues=bvalues\n",
    "} else {\n",
    "    missing_samples=rownames(pheno_data[!rownames(pheno_data) %in% colnames(bvalues),])\n",
    "    pheno_data=pheno_data[!rownames(pheno_data) %in% missing_samples,]\n",
    "    bvalues=bvalues[, colnames(bvalues) %in% rownames(pheno_data)]\n",
    "    paste0(\"The following samples were not found in the beta values but found in the pheno file: \", missing_samples)\n",
    "}\n",
    "\n",
    "# ordering pheno file based on beta\n",
    "paste0(\"Warning: this script orders the pheno file based on the beta matrix:\", Sys.time())\n",
    "pheno_data=pheno_data[match(rownames(pheno_data),colnames(bvalues)),]\n",
    "\n",
    "# transposing bvalues\n",
    "bvalues_t=t(bvalues)\n",
    "\n",
    "# building beta values structures\n",
    "fid_iid=data.frame(FID=rownames(bvalues_t),IID=rownames(bvalues_t))\n",
    "# adding fid_iid to bvalues matrices\n",
    "bvalues_t=cbind(fid_iid,bvalues_t)\n",
    "# saving the txt beta matrices for osca\n",
    "# making object for naming beta matrix\n",
    "betaout=paste0(outfile,\"_datafile_beta.txt\",sep=\"\")\n",
    "paste0(\"Start writting beta matrix\", Sys.time())\n",
    "# using data.table to optimized writting\n",
    "data.table::fwrite(bvalues_t, file=betaout, row.names=FALSE, quote=FALSE, sep=\"\\t\")\n",
    "\n",
    "# building phenofile structures\n",
    "pheno_f=pheno_data[opt$pheno]\n",
    "# adding fid_iid to phenofile\n",
    "pheno_f=cbind(fid_iid,pheno_f)\n",
    "# saving phenofile for osca\n",
    "# making object for output name of the pheno file\n",
    "phenoout=paste0(outfile,\"_datafile_pheno.txt\",sep=\"\")\n",
    "paste0(\"Start writting phenofile: \", Sys.time())\n",
    "write.table(pheno_f, file=phenoout, row.names=FALSE, quote=FALSE, sep=\"\\t\")\n",
    "\n",
    "# building covarfile structures\n",
    "covar_f=pheno_data[!colnames(pheno_data) %in% opt$pheno]\n",
    "# adding fid_iid to phenofile\n",
    "covar_f=cbind(fid_iid,covar_f)\n",
    "# saving covar for osca\n",
    "# making object for output name of the covar file\n",
    "covarout=paste0(outfile,\"_datafile_covar.txt\",sep=\"\")\n",
    "paste0(\"Start writting covarfile: \", Sys.time())\n",
    "write.table(covar_f, file=covarout, row.names=FALSE, quote=FALSE, sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fbb4f9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# executing the Rscript to transform data to glint format\n",
    "# move to the directory storing the Rscript\n",
    "Rscript qcdata_to_osca.Rscript --file=/vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/qced_data_v02062023.rds \\\n",
    "--out=/vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca/ \\\n",
    "--outname=qced_data_v02152023_osca \\\n",
    "--pheno=SES \\\n",
    "--covar=age,gender\n",
    "\n",
    "# deactivate environment\n",
    "conda deactivate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5ec1fb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# making osca files\n",
    "# setting osca path\n",
    "osca_p=\"/gpfs/gibbs/project/montalvo-ortiz/jjm262/programs/osca/osca-0.46.1-linux-x86_64/osca-0.46.1\"\n",
    "# setting beta matrix path\n",
    "beta_p=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca/qced_data_v02152023_osca_datafile_beta.txt\"\n",
    "# setting output\n",
    "out_p=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca/qced_data_v02152023_osca_files\"\n",
    "\n",
    "# running osca to build the osca files\n",
    "${osca_p} --efile ${beta_p} --methylation-beta --make-bod --out ${out_p}\n",
    "\n",
    "# updating annotation information\n",
    "# set annotation file\n",
    "annot_f=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/annot/annot_file.txt\"\n",
    "${osca_p} --befile ${out_p} --update-opi ${annot_f}\n",
    "\n",
    "# creating ORM\n",
    "# setting ORM output\n",
    "out_orm=\"/vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca/qced_data_v02152023_osca_orm\"\n",
    "${osca_p} --befile ${out_p} --make-orm --out ${out_orm}\n",
    "\n",
    "# splitting covar file in discrete and quantative covariates\n",
    "cd /vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca\n",
    "# cutting discrete covars\n",
    "cut -f 1-2,4 qced_data_v02152023_osca_datafile_covar.txt > qced_data_v02152023_osca_datafile_discovar.txt\n",
    "# cutting continuos covars\n",
    "cut --complement -f 4 qced_data_v02152023_osca_datafile_covar.txt > qced_data_v02152023_osca_datafile_contcovar.txt\n",
    "\n",
    "# running GREML\n",
    "${osca_p} --reml --orm ${out_orm} \\\n",
    "--pheno /vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca/qced_data_v02152023_osca_datafile_pheno.txt \\\n",
    "--covar /vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca/qced_data_v02152023_osca_datafile_discovar.txt \\\n",
    "--qcovar /vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca/qced_data_v02152023_osca_datafile_contcovar.txt \\\n",
    "--out /vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/results/assoc/mixed_models_osca/reml_all_samples\n",
    "\n",
    "# in this analysis we have an error\n",
    "# Performing  REML analysis ...\n",
    "# 134 observations, 10 fixed effect(s), and 2 variance component(s)(including residual variance).\n",
    "# Calculating prior values of variance components by EM-REML ...\n",
    "# Updated prior values: 735.682459        2136.204405\n",
    "# logL: -4931.300\n",
    "# Running AI-REML algorithm ...\n",
    "# Iter.   logL    V(O)    V(e)\n",
    "# Error: matrix H is not invertible.\n",
    "\n",
    "# running MOMENT model in OSCA\n",
    "${osca_p} --moment2-beta \\\n",
    "--befile ${out_p} \\\n",
    "--pheno /vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca/qced_data_v02152023_osca_datafile_pheno.txt \\\n",
    "--covar /vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/databases/qced/osca/qced_data_v02152023_osca_datafile_discovar.txt \\\n",
    "--moment-cor \\\n",
    "--cor-r2 0.6 \\\n",
    "--out /vast/palmer/scratch/montalvo-ortiz/jjm262/epigenomics/ewas_saliva_ses/results/assoc/mixed_models_osca/moment_all_samples_v02162023\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
